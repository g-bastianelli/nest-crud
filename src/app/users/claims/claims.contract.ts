import { initContract } from '@ts-rest/core';
import {
  paginationQuerySchema,
  paginationResponseSchema,
} from '../../utils/pagination.utils';
import { z } from 'zod';

const c = initContract();

const createClaimsRequestBodySchema = z.array(
  z.object({
    title: z.string(),
    description: z.string(),
    value: z.number(),
  }),
);

type CreateClaimsRequestBody = z.infer<typeof createClaimsRequestBodySchema>;

const userClaimsContract = c.router({
  listUserClaims: {
    method: 'GET',
    path: '/users/:userId/claims',
    pathParams: z.object({
      userId: z.string().uuid(),
    }),
    query: paginationQuerySchema,
    responses: {
      200: z.object({
        claims: z.array(
          z.object({
            id: z.string().uuid(),
            title: z.string(),
            description: z.string(),
            value: z.number(),
          }),
        ),
        pagination: paginationResponseSchema,
      }),
    },
    summary: "List user's claims",
  },
  batchCreateClaims: {
    method: 'POST',
    path: '/users/:userId/claims/batch',
    pathParams: z.object({
      userId: z.string().uuid(),
    }),
    body: createClaimsRequestBodySchema,
    responses: {
      201: z.object({
        claims: z.array(
          z.object({
            id: z.string().uuid(),
          }),
        ),
      }),
    },
    summary: 'Batch create claims for a user',
  },
  deleteUserClaim: {
    method: 'DELETE',
    path: '/users/:userId/claims/:claimId',
    pathParams: z.object({
      userId: z.string().uuid(),
      claimId: z.string().uuid(),
    }),
    responses: {
      204: z.void(),
    },
    summary: 'Delete a claim for a user',
  },
});

export { userClaimsContract, CreateClaimsRequestBody };
