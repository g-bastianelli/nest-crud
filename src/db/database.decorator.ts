import { Inject } from '@nestjs/common';
import { TEST_CRUD_DB } from './database.consts';

const InjectDatabase = () => Inject(TEST_CRUD_DB);

export { InjectDatabase };
