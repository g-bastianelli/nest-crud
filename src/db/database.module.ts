import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { drizzle } from 'drizzle-orm/node-postgres';
import * as schema from './database.schema';
import { DatabaseService } from './database.service';
import { TEST_CRUD_DB } from './database.consts';

@Module({
  providers: [
    {
      provide: TEST_CRUD_DB,
      inject: [ConfigService],
      useFactory: async (config: ConfigService) => {
        return drizzle(config.get('DATABASE_URL'), { schema });
      },
    },
    DatabaseService,
  ],
  exports: [TEST_CRUD_DB],
})
class DatabaseModule {}

export { DatabaseModule };
