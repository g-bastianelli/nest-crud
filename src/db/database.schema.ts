import { integer, pgTable, uuid, varchar } from 'drizzle-orm/pg-core';

export const users = pgTable('users', {
  id: uuid().defaultRandom().primaryKey(),
  name: varchar({ length: 255 }).notNull(),
  email: varchar({ length: 255 }).notNull().unique(),
});

export const claims = pgTable('claims', {
  id: uuid('id').defaultRandom().primaryKey().notNull(),
  userId: uuid()
    .references(() => users.id)
    .notNull(),
  title: varchar({ length: 255 }).notNull(),
  description: varchar({ length: 255 }).notNull(),
  value: integer().notNull(),
});
