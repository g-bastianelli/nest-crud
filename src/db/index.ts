export * from './database.consts';
export * from './database.decorator';
export * from './database.module';
export * from './database.types';

import * as schema from './database.schema';

export { schema };
