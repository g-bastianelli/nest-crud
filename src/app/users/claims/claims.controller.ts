import { Controller, Req } from '@nestjs/common';
import { tsRestHandler, TsRestHandler } from '@ts-rest/nest';
import { userClaimsContract } from './claims.contract';
import { UserClaimsService } from './claims.service';
import { buildPaginationResponse } from '../../utils/pagination.utils';

@Controller()
class UserClaimsController {
  constructor(private readonly userClaimsService: UserClaimsService) {}

  @TsRestHandler(userClaimsContract.listUserClaims)
  listUserClaims(@Req() request: Request) {
    return tsRestHandler(
      userClaimsContract.listUserClaims,
      async ({ params: { userId }, query: { page, pageSize } }) => {
        const { claims, total } = await this.userClaimsService.listClaims(
          userId,
          page,
          pageSize,
        );
        return {
          status: 200,
          body: {
            claims,
            pagination: buildPaginationResponse(request, {
              page,
              pageSize,
              total,
            }),
          },
        };
      },
    );
  }

  @TsRestHandler(userClaimsContract.batchCreateClaims)
  batchCreateClaims() {
    return tsRestHandler(
      userClaimsContract.batchCreateClaims,
      async ({ params: { userId }, body }) => {
        const claims = await this.userClaimsService.createClaims(userId, body);
        if (!claims) {
          return {
            status: 400,
            body: {
              message: 'User not found',
            },
          };
        }

        return {
          status: 200,
          body: claims,
        };
      },
    );
  }
}

export { UserClaimsController };
