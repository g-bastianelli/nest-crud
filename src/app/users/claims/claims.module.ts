import { Module } from '@nestjs/common';
import { UserClaimsService } from './claims.service';
import { UserClaimsController } from './claims.controller';
import { DatabaseModule } from '../../../db';

@Module({
  imports: [DatabaseModule],
  controllers: [UserClaimsController],
  providers: [UserClaimsService],
})
class UserClaimsModule {}

export { UserClaimsModule };
