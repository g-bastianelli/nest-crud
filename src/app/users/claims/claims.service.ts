import { Injectable } from '@nestjs/common';
import { eq, InferInsertModel, sql } from 'drizzle-orm';
import { CreateClaimsRequestBody } from './claims.contract';
import { DatabaseType, InjectDatabase, schema } from '../../../db';

@Injectable()
class UserClaimsService {
  constructor(@InjectDatabase() private readonly db: DatabaseType) {}

  async createClaims(userId: string, claimsToCreate: CreateClaimsRequestBody) {
    const user = await this.db.query.users.findFirst({
      where: (user, { eq }) => eq(user.id, userId),
    });
    if (!user) {
      return null;
    }

    return this.createClaimsQuery(
      claimsToCreate.map((claim) => ({ ...claim, userId })),
    );
  }

  async listClaims(userId: string, page: number, pageSize: number) {
    const [{ count }] = await this.countClaimsQuery(userId);

    const claims = await this.listClaimsQuery(userId, page, pageSize);

    return { claims, total: count };
  }

  private countClaimsQuery(userId: string) {
    return this.db
      .select({
        count: sql<number>`count(*)`,
      })
      .from(schema.claims)
      .where(eq(schema.claims.userId, userId));
  }

  private createClaimsQuery(claims: InferInsertModel<typeof schema.claims>[]) {
    return this.db.insert(schema.claims).values(claims).returning({
      id: schema.claims.id,
    });
  }

  private listClaimsQuery(userId: string, page: number, pageSize: number) {
    return this.db.query.claims.findMany({
      columns: {
        id: true,
        title: true,
        description: true,
        value: true,
      },
      where: (claim, { eq }) => eq(claim.userId, userId),
      limit: pageSize,
      offset: (page - 1) * pageSize,
      orderBy: (claim, { desc }) => desc(claim.value),
    });
  }
}

export { UserClaimsService };
