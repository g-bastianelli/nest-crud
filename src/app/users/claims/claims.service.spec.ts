import { describe } from 'node:test';
import { drizzle } from 'drizzle-orm/node-postgres';
import { schema } from '../../../db';
import { UserClaimsService } from './claims.service';

describe('UserClaimsService', () => {
  const userClaimsService = new UserClaimsService(drizzle('mock', { schema }));

  it('countClaimsQuery should generate the right query', function () {
    expect(
      userClaimsService['countClaimsQuery']('1').toSQL(),
    ).toMatchSnapshot();
  });

  it('listClaimsQuery should generate the right query', function () {
    expect(
      userClaimsService['listClaimsQuery']('1', 1, 10).toSQL(),
    ).toMatchSnapshot();
  });

  it('createClaimsQuery should generate the right query', function () {
    expect(
      userClaimsService['createClaimsQuery']([
        {
          userId: 'uuid',
          title: 'title',
          description: 'description',
          value: 1,
        },
      ]).toSQL(),
    ).toMatchSnapshot();
  });
});
