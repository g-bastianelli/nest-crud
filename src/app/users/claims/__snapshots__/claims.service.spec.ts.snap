// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countClaimsQuery should generate the right query 1`] = `
{
  "params": [
    "1",
  ],
  "sql": "select count(*) from "claims" where "claims"."userId" = $1",
}
`;

exports[`createClaimsQuery should generate the right query 1`] = `
{
  "params": [
    "uuid",
    "title",
    "description",
    1,
  ],
  "sql": "insert into "claims" ("id", "userId", "title", "description", "value") values (default, $1, $2, $3, $4) returning "id"",
}
`;

exports[`listClaimsQuery should generate the right query 1`] = `
{
  "params": [
    "1",
    10,
  ],
  "sql": "select "id", "title", "description", "value" from "claims" where "claims"."userId" = $1 order by "claims"."value" desc limit $2",
  "typings": [
    "uuid",
    "none",
  ],
}
`;
